### PR TITLE
Add examples to exportparts

### PR DESCRIPTION
**Pull request by KonnorRogers**
### Description

Adding examples for using exportparts

### Motivation

The exportparts documentation has no examples and simply links to the spec and is not very welcoming to new users.

### CODE REVIEW

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 1]
```diff
@@ -17,6 +17,76 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
 
+## Examples
+
+First let's create a card component that we will then wrap with another component.
+
+```js
+class CardComponent extends HTMLElement {
+  connectedCallback() {
+    this.attachShadow({ mode: "open" });
+    this.shadowRoot.innerHTML = `
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:28]: Let's use HTML to create HTML. This is a perfect opportunity to demonstrate using `<template>`

Comment (KonnorRogers) [reply files/en-us/web/html/global_attributes/exportparts/index.md:28]: I'm not sure I follow.

Something like this?

```html
<template id="card-component-template">
  <style>
    .base { display: grid };
  </style>
       
   <div class="base" part="base">
     <div part="header"><slot name="header"></slot></div>
     <div part="body"><slot></slot></div>
     <div part="footer"><slot name="footer"></slot></div>
    </div>
</template>
```

```js
class CardComponent extends HTMLElement {
  connectedCallback () {
    this.attachShadow({ mode: "open" })
    const firstChild = this.shadowRoot.firstElementChild
    const content = document.querySelector("#card-component-template").content.cloneNode(true)
    
   this.shadowRoot.innerHTML = ""
    
    this.shadowRoot.appendChild(content)
  }
}
```

Comment (estelle) [reply files/en-us/web/html/global_attributes/exportparts/index.md:28]: Yes. Exactly. Especially the HTML part.

For the HTML pages, where this example resides, I thought it's preferable to use the [`define`](https://developer.mozilla.org/en-US/docs/Web/API/Window/customElements/define) method of the Window's [customElements](https://developer.mozilla.org/en-US/docs/Web/API/Window/customElements) as it's [defined in the WHATWG spec](https://html.spec.whatwg.org/multipage/custom-elements.html#dom-window-customelements). 

```js
customElements.define('card-component',
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardComponent = document.getElementById('card-component-template').content;
      const shadowRoot = this.attachShadow({
        mode: 'open'
      });
      shadowRoot.appendChild(cardComponent.cloneNode(true));
    }
  });
  ```

example pen: https://codepen.io/estelle/pen/YzjNVjM

**Comment by bsmth** [comment]
Hi @KonnorRogers - thanks a lot for the contribution. There's some review comments pending for you to take a look at before this PR can progress, do you plan to come back to this? Thanks :)

**Comment by KonnorRogers** [comment]
@bsmth Totally forgot about this 🙈 I'll take a look and come back and fix it.

**Comment by bsmth** [comment]
> @bsmth Totally forgot about this 🙈 I'll take a look and come back and fix it.

No problem, I'm just checking in on PRs awaiting response to see if we can unblock them, thanks! :)

**Comment by KonnorRogers** [comment]
Updated!

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 2]
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
+
+Now we can target parts on the `<card-component>` from the `<card-wrapper>` like so:
+
+```css
+card-wrapper::part(header) {
+  font-weight: bold;
+}
+```
+
+### Renaming / Remapping parts
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:93]:
```suggestion
### Exposing mapped parts
```

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 3]
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
+
+To rename an exportpartm the syntax looks something like this: `exportparts="<original-name>:<new-name>,<original-name>:<new-name>"`
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:95]:
```suggestion
To rename exported parts, we include a comma-separated list of mapped parts, with each mapped part including the original name and exported name separated by a colon (`:`): 
```

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 4]
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
+
+And here's the prior `CardWrapper` with the remapping syntax:
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:97]:
```suggestion
#### HTML

We update the prior `<card-wrapper>` custom element with the remapping syntax (omitting `body` from the exported parts list):
```

we don't explain that `CardWrapper` is the same as `<card-wrapper>`, so let's keep it as an element.

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 5]
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
+    exportparts="
+       base:card__base, 
+       header:card__header, 
+       body:card__body, 
+       footer:card__footer
+     ">
+  </card-component>
+</template>
+```
+
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:111]:
```suggestion

#### JavaScript
```

also include the hidden JS and HTML

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 6]
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
+      const shadowRoot = this.attachShadow({
+        mode: "open",
+      });
+      shadowRoot.appendChild(cardWrapper.cloneNode(true));
+    }
+  },
+);
+```
+
+Now we can target parts on the `<card-component>` from the `<card-wrapper>` like so:
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:128]:
```suggestion
#### CSS

In targetting parts of the `<card-component>` from within the `<card-wrapper>`, we can only style the exported parts via their exposed part names:

```

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 7]
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
+
+```css
+card-wrapper::part(card__header) {
+  font-weight: bold;
+}
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:133]:
```suggestion
/* selects the exported parts name */
card-wrapper::part(card__header) {
  font-weight: bold;
}
/* selects nothing: these part names were not exported */
card-wrapper::part(footer),
::part(body) {
  font-weight: bold;
}
```

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 8]
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
+
+And here's the prior `CardWrapper` with the remapping syntax:
+
+```html
+<template id="card-wrapper">
+  <card-component
+    exportparts="
+       base:card__base, 
+       header:card__header, 
+       body:card__body,
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:105]:
```suggestion
```
Let's continue to omit body so we can show what happens when omitted.

**Review by estelle — COMMENTED** [review]
Thanks for all your work on this. This page was definitely missing an example, and this is a good one. It is coming along well! 

I made several suggestions to make this more understandable to people who don't know this stuff at all and to fit into how MDN writes examples

We should port some of this to the `parts` attribute page once this is complete; that page really needs help, and the intro explanation your wrote for this page will work well there.

**Comment by KonnorRogers** [comment]
@estelle thank you so much for the thorough review and the much more approachable phrasing!

I'll work on making these changes tomorrow!

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 9]
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
...
+<template id="card-component-template">
+  <style>
+    .base {
+      display: grid;
+    }
+  </style>
+  <div class="base" part="base">
+    <div part="header"><slot name="header"></slot></div>
+    <div part="body"><slot></slot></div>
+    <div part="footer"><slot name="footer"></slot></div>
```
[the remark this answers is not in the document]
Comment (KonnorRogers) [reply files/en-us/web/html/global_attributes/exportparts/index.md:41]: @estelle I did. I was using it as a "default" slot. But could see where not being explicit here could be confusing.

**Comment by KonnorRogers** [comment]
@estelle Sorry for taking a while, thank you so much for taking the time to make this content read so much nicer!

I applied your suggestions and removed the `` ```css`` at the end of that code block.

Let me know if there's anything else I can do to help push this across the finish line!

**Review by estelle — COMMENTED** [review]
i am going to add a commit to enable the other two examples to show.

**Comment by estelle** [comment]
I directly committed some code to show the results. I figured showing you how to do it would be faster than explaining it. , The results are showing, but they aren't what we were expecting. I am going to try troubleshooting in CodePen.

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 10]
```diff
@@ -30,6 +30,261 @@ When exporting a `part`, you have the option to assign a different name to the p
...
+
+Now we can target parts on the `<card-component>` from the `<card-wrapper>` like so:
+
+```css
+::part(body) {
+  color: red;
+  font-style: italic;
+}
+
+::part(header, footer) {
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:180]:
```suggestion
::part(header),
::part(footer) {
```

**Comment by estelle** [comment]
> Other than the slots not being populated on a nested component, those examples are working. Now i need to trouble shoot that.

I figured it out with the help of @OnkarRuikar. Thank you, Onkar!!!!

see https://codepen.io/estelle/pen/qBvKyEv

**Comment by estelle** [comment]
@KonnorRogers I updated the examples so that they are displayed and added the markup needed to nest the components. Are you OK with these changes?

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 11]
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
...
+
+::part(header),
+::part(footer) {
+  font-weight: bold;
+}
+```
+
+#### Results
+
+{{ EmbedLiveSample('Nested_component', '100%', '160') }}
```
Comment (estelle) [inline files/en-us/web/html/global_attributes/exportparts/index.md:207]:
```suggestion
{{ EmbedLiveSample('Nested_component', '100%', '400') }}
```

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 12]
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
...
+#### HTML
+
+First, let's create a card component that we will then wrap with another component. We also use the new element we created, populating the slots with plain text as content.
+
+```html
+<template id="card-component-template">
+  <style>
+    p {
+      margin: 0.5em 0;
+    }
```
Comment (OnkarRuikar) [inline files/en-us/web/html/global_attributes/exportparts/index.md:48]:
```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 13]
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
...
+
+#### HTML
+
+```html hidden
+<template id="card-component-template">
+  <style>
+    p,
+    h2 {
+      margin: 0.5em 0;
+    }
```
Comment (OnkarRuikar) [inline files/en-us/web/html/global_attributes/exportparts/index.md:113]:
```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 14]
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
...
+#### HTML
+
+We update the prior `<card-wrapper>` custom element with the remapping syntax (omitting `body` from the exported parts list):
+
+```html hidden
+<template id="card-component-template">
+  <style>
+    p {
+      margin: 0.5em 0;
+    }
```
Comment (OnkarRuikar) [inline files/en-us/web/html/global_attributes/exportparts/index.md:224]:
```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```

File: files/en-us/web/html/global_attributes/exportparts/index.md [hunk 15]
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
...
+</template>
+```
+
+```html
+<template id="card-wrapper">
+  <card-component exportparts="base, header, body">
+    <span slot="header_slot"><slot name="H"></slot></span>
+    <span slot="body_slot"><slot name="B"></slot></span>
+    <span slot="footer_slot"><slot name="F"></slot></span>
+  </card-component>
```
Comment (OnkarRuikar) [inline files/en-us/web/html/global_attributes/exportparts/index.md:129]:
```suggestion
  <style>
    .wrapper {
      padding: 1rem;
      border: 2px dashed red;
    }
  </style>
  <div class="wrapper">
    <card-component exportparts="base, header, body">
      <span slot="header_slot"><slot name="H"></slot></span>
      <span slot="body_slot"><slot name="B"></slot></span>
      <span slot="footer_slot"><slot name="F"></slot></span>
    </card-component>
  </div>
```

**Review by OnkarRuikar — COMMENTED** [review]
Would be great if we add a border to the card and wrapper. It would look like this:

File: files/en-us/web/html/global_attributes/exportparts/index.md (hunk 13 above)
Comment (estelle) [reply files/en-us/web/html/global_attributes/exportparts/index.md:112]: i added this to the component from the main DOM and added :host display block here instead; showing a few more examples of how to bypass the DOM / shadowDOM barrier

**Comment by KonnorRogers** [comment]
> I directly committed some code to show the results. I hope that is OK.
> 
> @KonnorRogers I updated the examples so that they are displayed and added the markup needed to nest the components. Are you OK with these changes?

Totally fine with me!! No issues on my end! I appreciate you taking the time to make this happen!

**Review by estelle — APPROVED** [review]
Yay! Thank you for making this happen. I learned so much working with you on this. 

Also, congratulations on your first merged MDN content PR. 
Welcome to the team!

**Comment by KonnorRogers** [comment]
❤️ Thank you @estelle so much for taking the time to flesh out my half baked PR and turn it into something much more meaningful 🙈  and thank you to everyone else who helped!

**Pull request merged**